fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/synchronizedsession/SynchronizedSession.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * SynchronizedSession.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb.synchronizedsession;
+
+import com.apple.foundationdb.MutationType;
+import com.apple.foundationdb.ReadTransaction;
+import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseRunner;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordContext;
+import com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer;
+import com.apple.foundationdb.subspace.Subspace;
+import com.apple.foundationdb.tuple.Tuple;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+@API(API.Status.EXPERIMENTAL)
+public class SynchronizedSession {
+    @Nonnull
+    private Subspace lockSubspace;
+    @Nonnull
+    private FDBDatabaseRunner runner;
+
+    @Nullable
+    private UUID sessionId = null;
+
+    // TODO: configurable?
+    private static final long LEASE_PERIOD_MILL = 60_000;
```

Comment:
Yes, definitely.

Reply:
@alecgrieser Yes. I raised it to @scgray before that this value could mean either something this session honers OR something this session wants others to honer. We agreed it should be the latter one. Looks like you agreed too.

As for keeping this information, there are two ways to go:
- One is saving the expiration time rather than the lock acquisition time as you suggested. The problem is that it can be possible that different workers on a same session have different settings for this parameter. Maybe this is not a problem because the max value will be accepted rather than the one written later. So I like this one.
- Another is persisting this parameter as the third part of lock information. I remember that was Scott's intent when we discussed last time. But I feel persisting one more part can make it hard to analyze all kinds of race conditions.

Reply:
Oh, I see: in theory, if you had different workers working on the same lock (e.g., different workers working on separate part of an index build all using the same lock) then you might have different workers updating the lease time. Yeah, I think because it uses `MAX`, whoever it will (effectively) use the setting of whoever has the highest configured value.